File: domain/src/commonMain/kotlin/com/voxenlabs/domain/server/usecases/StoreServerUseCase.kt
package com.voxenlabs.domain.server.usecases

import com.voxenlabs.domain.server.models.Server
import com.voxenlabs.domain.server.repositories.ServerRepositoryInterface
import org.koin.core.annotation.Factory

@Factory
class StoreServerUseCase(
    val serverRepository: ServerRepositoryInterface,
) {
    suspend operator fun invoke(server: Server) {
        if (!isServerAlreadyStored(server)) {
            serverRepository.storeServer(server)
        }
    }

    private suspend fun isServerAlreadyStored(server: Server): Boolean =
        serverRepository.getStoredServers().find { it.hostname == server.hostname && it.port == server.port } != null
}

File: src/modules/Voxen.Client.Domain/ServerDefinitions/UseCases/StoreServerUseCase.cs
using Voxen.Client.Domain.ServerDefinitions.Models;
using Voxen.Client.Domain.ServerDefinitions.RepositoryInterfaces;

namespace Voxen.Client.Domain.ServerDefinitions.UseCases;

public class StoreServerUseCase(IServerRepository serverRepository)
{
    private readonly IServerRepository serverRepository = serverRepository;

    public void Invoke(Server server)
    {
        if (!IsServerAlreadyStored(server))
        {
            serverRepository.StoreServer(server);
        }
    }

    private bool IsServerAlreadyStored(Server server)
    {
        return serverRepository.GetStoredServers().Contains(server);
    }
}
